db/commitlog/commitlog.cc

```diff
@@ -1259,10 +1259,22 @@ void db::commitlog::segment_manager::flush_segments(bool force) {
 
     // But if all segments are closed or we force-flush,
     // include all.
-    if (force || !active->is_still_allocating()) {
+    if (!active->is_still_allocating()) {
         high = replay_position(high.id + 1, 0);
     }
 
+    auto n = size_to_remove;
+
+    if (size_to_remove != 0) {
+        for (auto& s : _segments) {
```

Comment:
Won't this get called again and again? Or is it serialized against the previous call of flush_segments?

Reply:
I'm thinking we should consider only the last segment where we left off, towards the segment that marks the boundary where we want to flush.

Reply:
Once a segment is actually ready, it will be removed from this list. We will loop here every time until that time, but otoh we don't have to keep another state to keep track of relation "size we want to free" and "amount requested flushed". The list is usually ~10 or less, so should not be very expensive.

Reply:
Why is it 10 or less?

Reply:
Usually (in servers I run). It should be in the order of 10:s, since commitlog space is divided by shards, and we typically scale by sharding, not by throwing the same level of memory/disk to one shard. 
Point being, it is not a massive list. Esp not if we manage to get memtables to disk in an orderly fashion.

Reply:
i3 family has 7GB/shard. That's 218 segments. We can envision servers with more memory/shard.

Reply:
Even so, assuming we flush cur-max, i.e. less than half the size, you will loop through ~100 elems and do a subtract+check. Not expensive. 
The alternative is trying to keep state here. Which is problematic, because there is no absolute correlation between disk space used, segments made free by flush requests and segment order here. In fact, the database could at any time by it self elect to release CF/rp:s so that a segment on the "other" side of your clock handle is dropped, and we move back to acceptable disk footprint. But that does not mean that when we next want to remove X bytes should start from a state marker set lower. 
I still think simply collecting the X lowest RP segments making up the N bytes we want to remove each time is the cleanest way to do it. Can maybe shortcut it by keeping track of how much we've currently asked to remove and including this in the trigger so we don't repeatedly request the same flushes here.

Reply:
But I don't even think that is really necessary, since we already restrict the flush requests to when we've created new segments since the last - i.e. we've actually increased footprint, not just still are over it.

Reply:
How often to we iterate over all of these segments?

I'm for keeping this state, but after a major refactoring (and coroutinization) of the code. It's quite gnarly.

Reply:
Every time we've created a new segment since last check, and are/remain over threshold. Which means at worst, assuming instigated memtable flush did not help (yet), and an incoming data rate at 32MB/shard would mean once every second. 
Less with lower data rate.

Reply:
Not thrilled with it. Iterating over hundreds of segments can cause a small stall (probably won't exceed our task quota). But we can improve it after some refactoring makes it easier.